### PR TITLE
tools: Add capabilities and pattern for creating living usage examples and embedding them in documentation

### PR DESCRIPTION
**Pull request by Josmithr**
The ultimate goal here is to create a pattern for embedding code examples in our documentation in a way that offers confidence that the examples are up-to-date and correct.

There are 2 implementation aspects here:
1. Added functionality to `@fluid-tools/markdown-magic` to support embedding code-blocks.
2. Added an example pattern of creating sample usage code (via `@fluid-tools/api-markdown-documenter`), validating it during the package build, and embedding the usage code in the package's README.

### CODE REVIEW

File: tools/api-markdown-documenter/README.md [hunk 1]
```diff
@@ -68,23 +68,34 @@ Are you already generating `.api.json` report files as a part of your build?
 
 If yes, create a file called `api-markdown-documenter.js` and paste the following code:
 
-```javascript
-const { loadModel, MarkdownRenderer } = require("@fluid-tools/api-markdown-documenter");
+<!-- AUTO-GENERATED-CONTENT:START (INCLUDE_CODE:path=./examples/RenderMarkdown.ts&language=typescript&start=5) -->
```
Comment (Josmithr) [inline tools/api-markdown-documenter/README.md:71]: Sample output replacing the current, hard-coded (and _incorrect_) sample usage code with content directly embedded from a "live" code example that is build-validated.

File: tools/api-markdown-documenter/examples/RenderMarkdown.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
+/*!
```
Comment (Josmithr) [inline tools/api-markdown-documenter/examples/RenderMarkdown.ts:1]: Sample "live" code example.

File: tools/api-markdown-documenter/examples/tsconfig.json [hunk 3]
```diff
@@ -0,0 +1,12 @@
+{
```
Comment (Josmithr) [inline tools/api-markdown-documenter/examples/tsconfig.json:1]: There are a couple of important aspects to this `examples` config:
2. `paths` - We want our example code to look like _external_ usage would. So, we have aliased the `src` directory with the name of the package such that we can reference the code the same way end-users would.
1. `"noEmit": true` - we don't need to generate any javascript from this code; we strictly want to make sure it builds.

File: tools/markdown-magic/src/transforms/includeCodeTransform.cjs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+ * Default: <file-line-count> + 1.
+ * Constraints are the same as those for the `end` parameter to
+ * {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice#parameters | Array.slice}
+ * @param {string | undefined} options.language - The code language to use for syntax highlighting.
+ * E.g., "language=typescript" would yield a markdown codeblock starting with "\`\`\`typescript".
+ * Default: No language specified.
+ * @param {object} config - Transform configuration.
+ * @param {string} config.originalPath - Path to the document being modified.
+ */
+function includeCodeTransform(content, options, config) {
```
Comment (Josmithr) [inline tools/markdown-magic/src/transforms/includeCodeTransform.cjs:36]: `markdown-magic` functionality to support code-block embedding.

File: tools/api-markdown-documenter/examples/tsconfig.json (hunk 3 above)
Comment (alexvy86) [reply tools/api-markdown-documenter/examples/tsconfig.json:1]: An example that wanted to showcase importing from alpha/legacy/another-custom-entrypoint would just use that as the path key and still point to `../src`, right?

Comment (Josmithr) [reply tools/api-markdown-documenter/examples/tsconfig.json:1]: Confirmed: `..` works correctly for this. I've updated the config.

File: tools/markdown-magic/test/include-code.md
Comment (alexvy86) [inline tools/markdown-magic/test/include-code.md:1]: Remind me how these tests work? Is there something validating that this file is correct, or is it up to us to look at it and verify?

File: tools/markdown-magic/src/transforms/includeCodeTransform.cjs [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+ * @param {object} options - Transform options.
+ * @param {string} options.path - Relative path from the document to the file being embedded.
+ * @param {string | undefined} options.start - (optional) First line from the target file to be embedded (inclusive).
+ * Expected to be a string-formatted integer.
+ * Default: 0.
+ * Constraints are the same as those for the `end` parameter to
+ * {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice#parameters | Array.slice}
+ * @param {string | undefined} options.end - (optional) Line of the target file at which to end the embedded range (exclusive).
+ * Expected to be a string-formatted integer.
+ * Default: <file-line-count> + 1.
```
Comment (alexvy86) [inline tools/markdown-magic/src/transforms/includeCodeTransform.cjs:27]: I think this would be the correct default under the current logic (0-based indexing) but I actually like it being 1-indexed, thus `+ 1` here, but also start defaulting to 1.
```suggestion
 * Default: <file-line-count>.
```

File: tools/markdown-magic/src/transforms/includeCodeTransform.cjs [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+ * Constraints are the same as those for the `end` parameter to
+ * {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice#parameters | Array.slice}
+ * @param {string | undefined} options.language - The code language to use for syntax highlighting.
+ * E.g., "language=typescript" would yield a markdown codeblock starting with "\`\`\`typescript".
+ * Default: No language specified.
+ * @param {object} config - Transform configuration.
```
Comment (alexvy86) [inline tools/markdown-magic/src/transforms/includeCodeTransform.cjs:33]: Just curious, specifying the whole `config` in addition to the sub-properties works as expected? Even with this one in between some of its sub-properties?

File: tools/api-markdown-documenter/package.json [hunk 7]
```diff
@@ -24,10 +24,11 @@
 	},
 	"scripts": {
 		"api": "fluid-build . --task api",
-		"build": "concurrently npm:build:compile npm:lint && npm run build:readme && npm run build:docs",
+		"build": "concurrently npm:build:compile npm:lint && npm run build:examples && npm run build:readme && npm run build:docs",
```
[the remark this answers is not in the document]
Comment (Josmithr) [reply tools/api-markdown-documenter/package.json:27]: Yes - we need to build the examples _after_ we build the code, where it was built simultaneously before.

File: tools/markdown-magic/test/include-code.md
Comment (Josmithr) [reply tools/markdown-magic/test/include-code.md:1]: It's strictly PR based - if there are changes, the PR should verify that they make sense.

File: tools/markdown-magic/src/transforms/includeCodeTransform.cjs (hunk 5 above)
Comment (Josmithr) [reply tools/markdown-magic/src/transforms/includeCodeTransform.cjs:27]: I may actually rephrase this as "includes through the end of the file", rather than expressing in terms of indices.

File: tools/markdown-magic/src/transforms/includeCodeTransform.cjs (hunk 6 above)
Comment (Josmithr) [reply tools/markdown-magic/src/transforms/includeCodeTransform.cjs:33]: Can you elaborate? I'm not sure I understand the question.

File: tools/markdown-magic/src/utilities.cjs [hunk 8]
```diff
@@ -42,6 +38,29 @@ const readTemplate = (templateFileName, headingOffset = 0) => {
 	return unmodifiedContents.replace(/(^#)/gm, `$1${headingOffsetString}`);
 };
 
+
+/**
+ * Reads contents of the target file within the provided (optional) line boundaries.
```
[the remark this answers is not in the document]
Comment (Josmithr) [reply tools/markdown-magic/src/utilities.cjs:43]: I'd prefer to keep them 0-indexed. But I can be explicit about that.

File: tools/api-markdown-documenter/package.json (hunk 7 above)
[the remark this answers is not in the document]
Comment (alexvy86) [reply tools/api-markdown-documenter/package.json:27]: Oh, and is that because the examples try to import from the package itself? If they were importing from some external dependency we would be able to build source and examples in parallel? I'm fine with this approach, just wanted to understand the reason.

File: tools/markdown-magic/src/transforms/includeCodeTransform.cjs (hunk 6 above)
Comment (Josmithr) [reply tools/markdown-magic/src/transforms/includeCodeTransform.cjs:33]: Oh, in terms of the jsdoc comments? Yes, that is supported, and intellisense displays both.

Comment (alexvy86) [reply tools/markdown-magic/src/transforms/includeCodeTransform.cjs:33]: My bad, this is `config`, not `options` like above so there's no funky ordering. I guess my question would still be "do we need to specify `config` and `config.originalPath` separately" and if Intellisense does the right thing, but I imagine it does :).

File: tools/api-markdown-documenter/package.json (hunk 7 above)
[the remark this answers is not in the document]
Comment (Josmithr) [reply tools/api-markdown-documenter/package.json:27]: Yeah, because we're importing from the package itself.

File: tools/markdown-magic/src/transforms/includeCodeTransform.cjs (hunk 6 above)
Comment (Josmithr) [reply tools/markdown-magic/src/transforms/includeCodeTransform.cjs:33]: We don't strictly need the docs for `config`, the separate params are probably sufficient. It just felt incomplete to me without them.

Intellisense looks like this when hovering over the config:
image

**Review by alexvy86 — APPROVED** [review]

File: tools/markdown-magic/test/include-code.md [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+<!-- AUTO-GENERATED-CONTENT:END -->
+
+<!-- AUTO-GENERATED-CONTENT:START (INCLUDE_CODE:path=./sample-code.ts&language=typescript&start=6&end=-2) -->
+
+<!-- prettier-ignore-start -->
+<!-- NOTE: This section is automatically generated by embedding the referenced file contents. Do not update these generated contents directly. -->
+
+```typescript
+return 42;
+```
```
Comment (seanimam) [inline tools/markdown-magic/test/include-code.md:49]: Did you expect this code block to show up here? Seems out of place.

Comment (Josmithr) [reply tools/markdown-magic/test/include-code.md:49]: Yeah - the `INCLUDE_CODE` pragma will always introduce a code block. And the return statement was the line I was aiming for with `start=6&end=-2`
